### PR TITLE
Fixed #30757 -- Added a system check to ensure max_length fits the longest choice.

### PR DESCRIPTION
**Pull request by pope1ni**
Ticket [#30757](https://code.djangoproject.com/ticket/30757).

Following on from https://github.com/django/django/pull/11540#issuecomment-527511798.

### CODE REVIEW

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -270,16 +272,32 @@ def is_value(value, accept_promise=True):
                     for value, human_name in group_choices
                 ):
                     break
+                if self.max_length is not None:
+                    max_length = max(
+                        max_length,
+                        *(len(value) for value, human_name in group_choices if isinstance(value, str)),
+                    )
```
Comment (felixxm) [inline django/db/models/fields/__init__.py]: I think we can add check for `group_choices` to the `if` condition and simplify `max()`, e.g.
```python
if self.max_length is not None and group_choices:
    max_length = max(
        len(value) for value, _ in group_choices if isinstance(value, str)
    )
```

File: django/db/models/fields/__init__.py [hunk 2]
```diff
@@ -257,6 +257,8 @@ def is_value(value, accept_promise=True):
                 )
             ]
 
+        max_length = 0
```
Comment (felixxm) [inline django/db/models/fields/__init__.py]: Maybe `max_length` -> `choice_max_length`?

File: django/db/models/fields/__init__.py [hunk 3]
```diff
@@ -270,16 +272,32 @@ def is_value(value, accept_promise=True):
...
+                        max_length,
+                        *(len(value) for value, human_name in group_choices if isinstance(value, str)),
+                    )
             except (TypeError, ValueError):
                 # No groups, choices in the form [value, display]
                 value, human_name = group_name, group_choices
                 if not is_value(value) or not is_value(human_name):
                     break
+                if self.max_length is not None and isinstance(value, str):
+                    max_length = max(max_length, len(value))
```
Comment (felixxm) [inline django/db/models/fields/__init__.py]: Why we use `max()` here? :thinking: i.e.
```python
max_length = len(value)
```
works fine for me.

File: docs/ref/checks.txt [hunk 4]
```diff
@@ -145,6 +145,8 @@ Model fields
 * **fields.E121**: ``max_length`` must be a positive integer.
 * **fields.W122**: ``max_length`` is ignored when used with
   ``<integer field type>``.
+* **fields.E123**: ``max_length`` is too small to fit the largest value in
```
Comment (felixxm) [inline docs/ref/checks.txt]: Maybe `longest` instead of `largest`

File: django/db/models/fields/__init__.py (hunk 3 above)
Comment (pope1ni) [reply django/db/models/fields/__init__.py]: Don't forget that there is a loop `for choices_group in self.choices:` and we need the longest value, not the length of the last value.

File: django/db/models/fields/__init__.py (hunk 1 above)
Comment (pope1ni) [reply django/db/models/fields/__init__.py]: Same again here. (So yes to the `if` condition, but no to the simplification. Well I'll make it better, but we need the `max_choices_length = max(max_choices_length, *(...))` pattern.)

**Comment by pope1ni** [comment]
Hi @felixxm I think this should be good now.

File: tests/invalid_models_tests/test_ordinary_fields.py [hunk 5]
```diff
@@ -304,6 +304,36 @@ class Model(models.Model):
...
+            with self.subTest(name):
+                field = Model._meta.get_field(name)
+                self.assertEqual(field.check(), [
+                    Error(
+                        "'max_length' is too small to fit the longest value in "
+                        "'choices' (%d characters)." % choice_max_length,
+                        obj=field,
+                        id='fields.E123',
+                    ),
+                ])
```
Comment (felixxm) [inline tests/invalid_models_tests/test_ordinary_fields.py]: `max_length` and `choices` can be set also for other fields, e.g. `TextField` or `EmailField`, that's why I'm going to move this check to the `fields.E009`.

**Pull request merged**